core/shared/src/main/scala/scalaz/zio/internal/OneShot.scala

```diff
@@ -0,0 +1,43 @@
+// Copyright (C) 2018 John A. De Goes
+package scalaz.zio.internal
+
+/**
+ * A variable that can be set a single time. The synchronous,
+ * effectful equivalent of `Promise`.
+ */
+private[zio] class OneShot[A] private (@volatile var value: A) {
+
+  /**
+   * Sets the variable to the value. The behavior of this function
+   * is undefined if the variable has already been set.
+   */
+  final def set(v: A): Unit =
+    this.synchronized {
+      if (value != null) throw new Error("Defect: OneShot being set twice")
+
+      value = v
+
+      this.notifyAll()
+    }
+
+  /**
+   * Retrieves the value of the variable, blocking if necessary.
+   */
+  final def get: A = {
+    while (value == null) {
+      this.synchronized {
+        if (value == null) this.wait()
```

Comment:
I've seen this expressed like so:
```scala
this.synchronized {
  while (value == null) {
    this.wait()
  }
}
```

Reply:
That pattern will _always_ lock. It makes sense to short-circuit and lock only if necessary (if `value != null` then there's no point in acquiring the lock).